EXTENSIONS.md

```diff
@@ -0,0 +1,236 @@
+- Second parameter: a context object `{ kind: "let" }`
+- Return value: A `{ get, set, value }` object, with potentially new behavior
+
+This example can be desugared as:
+
+```js
+let { get_x, set_x, value: _x } = deprecated({value: 1, get() { return _x; }, set(v) { _x = v; }}, {kind: "let"});
+
+set_x(get_x()++);
+```
+
+## `const` decorators
+
+Variables declared with `const` can be decorated more simply--the decorator simply wraps the value being decorated when it's being defined.
+
+```js
+function inc(x) { return x+1; }
+const @inc x = 1;  // 2
+```
+
+`const` decorator details:
+- First parameter: The value of the RHS
+- Second parameter: a context object `{ kind: "const" }`
+- Return value: A new value for the variable
+
+This could be desugared as follows:
+
+```js
+const x = inc(1, {kind: "const"});
+```
+
+(This form isn't so useful by itself, but may become more important if future proposals share more information through the context object.)
+
+## Object literal and property decorators and annotations
+
+- A decorated object literal works like a class decorator, but with `kind: "object"`.
+- A decorated method, getter or setter in an object literal works just like one in a class, replacing that method.
+- A decorated object property works like a field decorator, but with `kind: "property"`, and it receives the initial value in the `value` property of the input, and returns it in the output object, rather than returning an initializer function, since it only runs once (in this way, it is similar to `let` decorators).
+
+Example:
+
+```js
+const x = @decA {
+  @decB p: v,
+  @decC m() { }
+};
+```
+
+<!--
+would desugar roughly into (in terms of the [`private`/`with` proposal](https://gist.github.com/littledan/5451d6426a8ed65c0f3c2822c51314d1))
+
+```js
+private #p;
+let {get, set, value} = decB() // tood: fihish)
+
+const x = decA({
+  #p: initializer(value)
+  valuedec
+  m: decC(function m() { }, {kind: "method"})
+}, {kind: "object"});
+```
+-->
+
+## Block decorators
+
+Decorating a block could wrap it as a function.
+
+```js
+@foo { bar; }
+```
+
+could desugar to
+
+```js
+foo(() => bar, { kind: "block" });
+```
+
+Similarly, in a class:
+
+```js
+class F {
+  @foo { bar; }
+}
+```
+
+could be equivalent to
+
+```js
+class F {
+  #_() { bar; }
+  constructor() {
+    foo.call(this, this.#_, {kind: "class-block"});
+  }
+}
+```
+
+These patterns might improving ergonomics with "hooks"-like patterns and component lifecycle methods.
+
+Note that this syntax could only work if it's in statement context; otherwise, it would be decorating an object literal.
+
+## Initializer decorators
+
+Decorating an initializer would turn the initializer in a thunk, so it can be run in the appropriate context (e.g., with usage tracking on) or re-run later (e.g., when dependencies change).
+
+```js
+let x @foo = bar;
+```
+
+could desugar into
+
+```js
+let x = foo(() => bar, {kind: "initializer"});
```

Comment:
How would this be morally any different from just `let x = foo(bar)` in most general uses?

Reply:
It lets you evaluate the initializer multiple times.

Reply:
It lets you evaluate the initializer multiple times.

Reply:
It lets you evaluate the initializer multiple times.

Reply:
It lets you evaluate the initializer multiple times.